lib/Date/Utility.pm

```diff
@@ -840,6 +838,35 @@ sub months_ahead {
     return month_number_to_abbrev($new_month) . '-' . $new_year;
 }
 
+=head2 move_to_nth_dow
+
+Takes an integer as an ordinal and a day of week representation
+
+The following are all equivalent:
+C<move_to_nth_dow(3, 'Monday')>
+C<move_to_nth_dow(3, 'Mon')>
+C<move_to_nth_dow(3, 1)>
+
+Returning the 3rd Monday of the month represented by the object or
+C<undef> if it does not exist.
+
+An exception is thrown on improper day of week representations.
+
+=cut
+
+sub move_to_nth_dow {
+    my ($self, $nth, $dow_abb) = @_;
+
+    $dow_abb //= 'undef';    # For nicer error reporting below.
+
+    my $dow = $days_to_num{lc $dow_abb} // croak 'Invalid day of week. We got [' . $dow_abb . ']';
+
+    my $dow_first = ($self->day_of_month - $self->day_of_week) % 7;
+    my $dom = ($dow + 7 - $dow_first) % 7 + ($nth - 1) * 7 + 1;
+
+    return try { Date::Utility->new(join '-', $self->year, $self->month, $dom) };
+}
```

Comment:
Ok, I ran your script with `2e3-1` replacing `1000` and a `for` loop of `1..15` replacing `19`:

```
ok 29986 - 4. Monday of Jun 2020 is 2020-06-22
ok 29987 - 4. Monday of Jun 2020 is 2020-06-22
ok 29988 - 4. Monday of Jun 2020 is 2020-06-22
ok 29989 - 4. Monday of Jun 2020 is 2020-06-22
ok 29990 - 4. Monday of Jun 2020 is 2020-06-22
ok 29991 - 4. Monday of Jun 2020 is 2020-06-22
ok 29992 - 4. Monday of Jun 2020 is 2020-06-22
ok 29993 - 4. Monday of Jun 2020 is 2020-06-22
ok 29994 - 4. Monday of Jun 2020 is 2020-06-22
ok 29995 - 4. Monday of Jun 2020 is 2020-06-22
ok 29996 - 4. Monday of Jun 2020 is 2020-06-22
ok 29997 - 4. Monday of Jun 2020 is 2020-06-22
ok 29998 - 4. Monday of Jun 2020 is 2020-06-22
ok 29999 - 4. Monday of Jun 2020 is 2020-06-22
ok 30000 - 4. Monday of Jun 2020 is 2020-06-22
1..30000
```

So, this seems to be working now.  The algorithm, as written, seems too complex because it matches the way I thought about solving it.  I am going to enshrine a smaller version of the test in the suite and work on a simplification.  Thanks for the note.. my extensive testing was along the wrong axis to find the issue!